crates/data_components/src/dynamodb/stream.rs

```diff
@@ -0,0 +1,603 @@
+                "data" => {
+                    let data_struct_builder = downcast_builder::<StructBuilder>(field_builder)
+                        .map_err(|e| StreamError::DeserializationError {
+                            message: e.to_string(),
+                        })?;
+                    append_item_to_struct_builder(&item_data, data_struct_builder).map_err(
+                        |e| StreamError::DeserializationError {
+                            message: e.to_string(),
+                        },
+                    )?;
+                }
+                _ => unreachable!("Unexpected field in changes schema {}", field.name()),
+            }
+        }
+    }
+
+    let struct_array = changes_struct_builder.finish();
+    let record_batch: RecordBatch = struct_array.into();
+
+    let change_batch =
+        ChangeBatch::try_new(record_batch).map_err(|e| StreamError::DeserializationError {
+            message: e.to_string(),
+        })?;
+
+    Ok(ChangeEnvelope::new(
+        Box::new(DynamoDBStreamCommitter::new()),
+        change_batch,
+    ))
+}
+
+fn streams_to_dynamodb_item(
+    item: HashMap<String, StreamsAttributeValue>,
+) -> HashMap<String, DynamoDbAttributeValue> {
+    item.into_iter()
+        .map(|(k, v)| (k, streams_to_dynamodb_attribute(&v)))
+        .collect()
+}
+
+fn streams_to_dynamodb_attribute(value: &StreamsAttributeValue) -> DynamoDbAttributeValue {
+    match value {
+        StreamsAttributeValue::B(blob) => DynamoDbAttributeValue::B(blob.clone()),
+        StreamsAttributeValue::Bool(b) => DynamoDbAttributeValue::Bool(*b),
+        StreamsAttributeValue::Bs(blobs) => DynamoDbAttributeValue::Bs(blobs.clone()),
+        StreamsAttributeValue::L(list) => {
+            DynamoDbAttributeValue::L(list.iter().map(streams_to_dynamodb_attribute).collect())
+        }
+        StreamsAttributeValue::M(map) => DynamoDbAttributeValue::M(
+            map.iter()
+                .map(|(k, v)| (k.clone(), streams_to_dynamodb_attribute(v)))
+                .collect(),
+        ),
+        StreamsAttributeValue::N(n) => DynamoDbAttributeValue::N(n.clone()),
+        StreamsAttributeValue::Ns(ns) => DynamoDbAttributeValue::Ns(ns.clone()),
+        StreamsAttributeValue::Null(n) => DynamoDbAttributeValue::Null(*n),
+        StreamsAttributeValue::S(s) => DynamoDbAttributeValue::S(s.clone()),
+        StreamsAttributeValue::Ss(ss) => DynamoDbAttributeValue::Ss(ss.clone()),
+        _ => DynamoDbAttributeValue::Null(true),
+    }
+}
+
+fn downcast_builder<T: ArrayBuilder>(builder: &mut dyn ArrayBuilder) -> Result<&mut T> {
+    let builder = builder
+        .as_any_mut()
+        .downcast_mut::<T>()
+        .context(DowncastBuilderSnafu)?;
+    Ok(builder)
+}
+
+struct DynamoDBStreamCommitter;
+
+impl DynamoDBStreamCommitter {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl CommitChange for DynamoDBStreamCommitter {
+    fn commit(&self) -> std::result::Result<(), CommitError> {
```

Comment:
DynamoDB has no way for clients to indicate where they are in the stream? How does Spice recover if it dies/restarts in the middle of processing?